core/src/main/java/hudson/model/queue/WorkUnitContext.java

```diff
@@ -108,12 +108,17 @@ public WorkUnit getPrimaryWorkUnit() {
      * All the {@link Executor}s that jointly execute a {@link Task} call this method to synchronize on the start.
      */
     public void synchronizeStart() throws InterruptedException {
```

Comment:
I think we want to set the exception state of the future for exceptional exits.

Also we can determine if this is the main work unit up front as none of the ISE exceptions in my suggestion will ever be thrown by the only internal code path
```suggestion
    /**
     * All the {@link Executor}s that jointly execute a {@link Task} call this method to synchronize on the start.
     */
    @Restricted(NoExternalUse.class)
    public void synchronizeStart() throws InterruptedException {
        Executor e = Executor.currentExecutor();
        if (e == null) {
            throw new IllegalStateException("synchronizeStart shall only be called from an Executor");
        }
        WorkUnit wu = e.getCurrentWorkUnit();
        if (wu == null) {
            throw new IllegalStateException("synchronizeStart shall only be called after a WorkUnit has been assigned");
        }
        boolean mainWork = wu.isMainWork();
        try {
            startLatch.synchronize();
            // the main thread will send a notification
            if (mainWork) {
                future.start.set(e.getCurrentExecutable());
            }
        } catch (Error | RuntimeException | InterruptedException problem) {
            if (mainWork) {
                future.start.set(problem);
            }
            throw problem;
        } finally {
            if (mainWork && !future.start.isDone()) {
                future.start.set(e.getCurrentExecutable());
            }
        }
    }
```